### ISSUE TITLE
Investigate Mastery Cache

### ISSUE THREAD

**Issue by seanpeters86** [WarcraftPriests/sl-shadow-priest#197]
Right now the simc module does some pretty expensive checks on each action to see what dots are currently ticking. We should see if adding some sort of cache to this makes things perform better.

**Comment by seanpeters86** [comment]
Tried out some stuff here: https://github.com/simulationcraft/simc/compare/priest-mastery-cache

Current Notes:
- actually slows the sim down :/
- Not sure if `last_tick` is the correct place to invalidate the cache
- Need to clean up the location of the cache struct so we don't have to make it mutable. I think this should be possible but not sure. When I tried it was very upset that everything was a const.
- Currently a very small DPS loss, looks to be because of where/when we invalidate the cache doesn't trigger the check on that first hit (which should be affected by mastery)
